sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala

```diff
@@ -148,3 +148,106 @@ private[joins] object HashedRelation {
     }
   }
 }
+
+
+/**
+ * A HashedRelation for UnsafeRow, which is backed by BytesToBytesMap that maps the key into a
+ * sequence of values.
+ *
+ * TODO(davies): use BytesToBytesMap
+ */
+private[joins] final class UnsafeHashedRelation(
+    private var hashTable: JavaHashMap[UnsafeRow, CompactBuffer[UnsafeRow]],
+    private var rowTypes: Array[DataType])
+  extends HashedRelation with Externalizable {
+
+  def this() = this(null, null)  // Needed for serialization
+
+  // UnsafeProjection is not thread safe
+  @transient lazy val fromUnsafeProjection = new ThreadLocal[FromUnsafeProjection]
```

Comment:
Ah, gotcha; that makes perfect since.  When you update this PR, could you maybe add a one-line comment to the class-level Scaladoc to explain that it's thread safe and thus is safe to store in a broadcast variable?